### DISCUSSION TITLE
Performance impact

### DISCUSSION THREAD

**Discussion by MalacTheLittle** [airsdk/Adobe-Runtime-Support#1914] [Q&A]
I need help choosing aproach that has better performance impact (if there is any difference in following example).
Let's imagine you have controler for creating and resizing objects on stage.

---
**FIRST approach** would be:
Create Class named "Controler" with following code:
```
for (var i:uint = 1, l:uint = 1000; i<=l; ++i){
	var item:MovieClip = new Item();
	item.resize(500, 500);
}
```
Create Class named "Item" with following method:
```
function resize(_width:uint, _height:uint):void{
	this.width = _width;
	this.height = _height;
}
```
---
**SECOND approach** would be:
Create Class named "Controler" with following code:
```
for (var i:uint = 1, l:uint = 1000; i<=l; ++i){
	var item:MovieClip = new Item();
	resize(item, 500, 500);
}

function resize(item:MovieClip, _width:uint, _height:uint):void{
	item.width = _width;
	item.height = _height;
}
```
Create Class named "Item" without any code in it.

---
Now, let's assume item has much more code and methods in it. 
Is there any difference in performance between these two approaches and does first approach uses more memory/processor?

**Comment by ajwfrost** [comment]
Hi

My first instinct was to say the first approach would be more efficient. If you try these two approaches out and look at the resulting ActionScript byte code that's generated (via the SWF Investigator decompiler perhaps) then you would see that there's some extra cost to the "item.width" property set in the second approach: you first have to retrieve the "item" object and then set the "width" property on it. 

Whereas with the first approach the "this" object is always available.

Whether that difference actually makes much difference after it's JIT-compiled, I'm not so sure; but also, the second approach could be improved by using "with item" (https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/statements.html#with). I was going to suggest also making that function inline which would help remove the initial cost of calling it within that loop... but I don't think this qualifies for inlining (not sure why but this seems to have some restrictions to it..)

So I'm not sure there would be a huge amount in it; the issues may be more relating to what properties/methods you're then calling on them. Memory-wise there's also not likely to be a big difference, function calls take up a bit of memory on the call stack so reducing these may help a little (again, inlining if possible) but not a massive change I would think.

I'd be interested to see if you get any empirical evidence of a different though :-)

thanks

**Reply by MalacTheLittle** [reply]
Thanks for taking time for this thread :)

I find first approach also more elegant from developers point of view since code is separated and easier to read/maintain/reuse, but the reason I started this thread is that I wasn't sure what's going on behind the scene. 
The case is that I'm making a menu with a large number of items that are instantiated/controled by menu itself and these items has other items that are instantiated by items itself. 
So I was wondering if instantiation of item takes more memory and more time (performance) if item has more code in it, but obviously you gain some and loose some with both approaches and it's acctually not much difference in my case :)

I was completely not aware of the `with` statement, it's total discovery for me, I can see a lot of use cases for it! Thanks for that :)

**Comment by 2jfw** [comment]
The following result with *1.000.000* iterations:

**Debug Build:**
Benchmark 1,  Time in ms: **4898**
Benchmark 2,  Time in ms: **5479**


**Release Build:**
Benchmark 1,  Time in ms: **3695**
Benchmark 2,  Time in ms: **4164**

---

The following result with *10.000* iterations:

**Debug Build:**
Benchmark 1,  Time in ms: **48**
Benchmark 2,  Time in ms: **64**

**Release Build:**
Benchmark 1,  Time in ms: **29**
Benchmark 2,  Time in ms: **36**

---


**Note:**
- Every "Run" will have slightly different outputs
- Performance also based on your local machine's performance
- Due to the high number of iterations it can take a couple of seconds for the application to show up


**Source Code:**

	package
	{
	
		import flash.display.Sprite;

		import flash.text.TextField;
		import flash.utils.getTimer;
	
	
		public class Main extends Sprite
		{
			private var _loggingTextField : TextField;
			private var _startTime : Number;
			private var _endTime : Number;
			private var _numIterations : int = 1000000;
	
	
			public function Main()
			{
				setupLoggingTextfield();
	
				startBenchmark1();
				startBenchmark2();
			}
	
	
			private function setupLoggingTextfield() : void
			{
				_loggingTextField            = new TextField();
				_loggingTextField.width      = stage.stageWidth;
				_loggingTextField.height     = stage.stageHeight;
				_loggingTextField.selectable = true;
	
				addChild(_loggingTextField);
			}
	
	
			private function startBenchmark1() : void
			{
				var item : Item; // outside of loop based on https://cwiki.apache.org/confluence/display/FLEX/Flex+SDK+coding+conventions+and+best+practices
				var len : int = _numIterations; // convert a to a local var to speed up access time in this scope
	
				_startTime = getTimer();
	
				for (var i : int = 0; i < len; i++)
				{
					item = new Item();
					item.resize(500,
					            500);
				}
	
				_endTime = getTimer();
	
				logTime("1");
			}
	
	
			private function startBenchmark2() : void
			{
				var item : Item; // outside of loop based on https://cwiki.apache.org/confluence/display/FLEX/Flex+SDK+coding+conventions+and+best+practices
				var len : int = _numIterations; // convert a to a local var to speed up access time in this scope
	
				_startTime = getTimer();
	
				for (var i : int = 0; i < len; i++)
				{
					item = new Item();
					resizeItem(item,
					           500,
					           500);
				}
	
				_endTime = getTimer();
	
				logTime("2");
			}
	
	
			private function resizeItem(item : Item,
			                            width : Number,
			                            height : Number) : void
			{
				item.width  = width;
				item.height = height;
			}
	
	
			private function logTime(id : String) : void
			{
				_loggingTextField.appendText("Benchmark " + id + ",  Time in ms: " + (_endTime - _startTime) + "\n");
			}
		}
	}

---



	package
	{
		  import flash.display.MovieClip;
	  
	  
		  public class Item extends MovieClip
		  {
			  public function Item()
			  {
			  }
	  
	  
			  public function resize(width : uint,
			                         height : uint) : void
			  {
				  this.width  = width;
				  this.height = height;
			  }
		  }
	  }

---

**Conclusion:**
Difference only "noticeable" on many (!) items


*EDIT:* Added test for 10k items and did some code polish


[PerformanceImpact#1914.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/8700805/PerformanceImpact.1914.zip)

**Reply by ajwfrost** [reply]
Interesting! So a reasonable difference if you do the "item.width" and "item.height" setting, wondering if it changes with the use of `with`.. :-)

**Reply by 2jfw** [reply]
Test with `with` (Bechmark 3) and 1.000.000 iterations:

**Debug Version:** Run 1
Benchmark 1,  Time in ms: **4199**
Benchmark 2,  Time in ms: **4162**
Benchmark 3,  Time in ms: **4270**

**Debug Version:** Run 2
Benchmark 1,  Time in ms: **5163**
Benchmark 2,  Time in ms: **5448**
Benchmark 3,  Time in ms: **6711**

**Debug Version:** Run 3
Benchmark 1,  Time in ms: **4642**
Benchmark 2,  Time in ms: **5514**
Benchmark 3,  Time in ms: **6489**

---

**Release Version:** Run 1
Benchmark 1,  Time in ms: **2941**
Benchmark 2,  Time in ms: **3606**
Benchmark 3,  Time in ms: **4074**

**Release Version:** Run 2
Benchmark 1,  Time in ms: **2958**
Benchmark 2,  Time in ms: **2767**
Benchmark 3,  Time in ms: **3010**

**Release Version:** Run 3
Benchmark 1,  Time in ms: **2918**
Benchmark 2,  Time in ms: **3615**
Benchmark 3,  Time in ms: **3929**

---
**Conclusion:**
`with` Statement seems to be slowest approach..

**Reply by 2jfw** [reply]
Oh, had small issue in the `with` function by assigning the `width` and `height` from the item itselt and not from the function parameters but it looks like using the function params is even slower.

Wrong version was:
 
		private function resizeItem2(item : Item,
		                             _width : Number,
		                             _height : Number) : void
		{
			with (item)
			{
				width  = width;
				height = height;
			}
		}


Correct version is

		private function resizeItem2(item : Item,
		                             _width : Number,
		                             _height : Number) : void
		{
			with (item)
			{
				width  = _width;
				height = _height;
			}
		}

---
**Release Version: Run 1**
Benchmark 1,  Time in ms: **3008**
Benchmark 2,  Time in ms: **3602**
Benchmark 3,  Time in ms: **4501**

**Release Version: Run 2**
Benchmark 1,  Time in ms: **3271**
Benchmark 2,  Time in ms: **3896**
Benchmark 3,  Time in ms: **3865**

**Release Version: Run 3**
Benchmark 1,  Time in ms: **3287**
Benchmark 2,  Time in ms: **4000**
Benchmark 3,  Time in ms: **5038**

**Release Version: Run 4**
Benchmark 1,  Time in ms: **3328**
Benchmark 2,  Time in ms: **4166**
Benchmark 3,  Time in ms: **5412**

**Reply by 2jfw** [reply]
**Debug Version: Run 1**
Benchmark 1,  Time in ms: **4534**
Benchmark 2,  Time in ms: **5180**
Benchmark 3,  Time in ms: **6490**

**Debug Version: Run 2**
Benchmark 1,  Time in ms: **4541**
Benchmark 2,  Time in ms: **5261**
Benchmark 3,  Time in ms: **6644**

**Debug Version: Run 3**
Benchmark 1,  Time in ms: **4670**
Benchmark 2,  Time in ms: **5108**
Benchmark 3,  Time in ms: **6747**

**Reply by ajwfrost** [reply]
Interesting. Not quite what I'd have expected! but there's a clear winner :-)

**Reply by MalacTheLittle** [reply]
@2jfw Wow, thanks for taking time to make these benchmarks! 
Now it's clear the first approach is not only cleaner from developers point of view but also from performance side. And `with` statement is to be used only in case to have cleaner code :)

Thanks!

**Reply by 2jfw** [reply]
You're welcome! It only took a couple of minutes to do so and it is always interesting to see what the best performance is :)

**Comment by 2jfw** [comment]
For completeness: Updated Benchmark sources (with `with` statement and a slightly different approach to Benchmark 1):
[PerformanceImpact#1914.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/8706542/PerformanceImpact.1914.zip)

**Reply by MalacTheLittle** [reply]
Interesting! I would thought that 1 and 1a would be the same... Could that be up to code parsing?

**Reply by 2jfw** [reply]
> Interesting! I would thought that 1 and 1a would be the same... Could that be up to code parsing?

I would have thought that **1a** would actually be even faster but I think that a `Function`, when it is returning a value, is the bottleneck here.

		private function startBenchmark1a() : void
		{
			var item : Item; // outside of loop based on https://cwiki.apache.org/confluence/display/FLEX/Flex+SDK+coding+conventions+and+best+practices
			var len : int = NUM_ITERATIONS; // convert a to a local var to speed up access time in this scope

			var startTime : Number = getTimer();

			for (var i : int = 0; i < len; i++)
			{
				item = new Item().resize2(500,
				                          500);
			}


			logTime("1a",
			        getTimer() - startTime);
		}

---

		public function resize2(width : uint,
		                        height : uint) : Item
		{
			this.width  = width;
			this.height = height;

			return this;
		}

---
3x Release Build Runs:

Benchmark 1,  Time in ms: **3802**
Benchmark 1a,  Time in ms: **4452**
Benchmark 2,  Time in ms: **4531**
Benchmark 3,  Time in ms: **4820**

Benchmark 1,  Time in ms: **3380**
Benchmark 1a,  Time in ms: **4430**
Benchmark 2,  Time in ms: **4271**
Benchmark 3,  Time in ms: **5679**

Benchmark 1,  Time in ms: **3486**
Benchmark 1a,  Time in ms: **3726**
Benchmark 2,  Time in ms: **3171**
Benchmark 3,  Time in ms: **4139**
